File: framework/languages/dotnet/e2e/RegistrationCodec/Server/InvalidDuplicate/RegistrationHandlers.cs
using RegistrationCodec.Server.Infrastructure;
using RegistrationCodec.Server.InvalidDuplicate.Infrastructure;
using RegistrationCodec.Shared;
using Zlink.Framework.Contracts.Handlers;

namespace RegistrationCodec.Server.InvalidDuplicate.Handlers;

[ZLinkHandlerGroup("auto")]
internal sealed class EchoAutoRequestHandler(EvidenceStore evidence)
    : IZLinkRequestHandler<EchoAutoReq, EchoReply>
{
    public ValueTask<EchoReply> HandleAsync(EchoAutoReq request, ZLinkRequestContext context, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add($"echo-request|variant=auto|value={request.Value}|content={context.ContentType}");
        return ValueTask.FromResult(new EchoReply($"echo:{request.Value}", context.ContentType ?? "<null>"));
    }
}

[ZLinkHandlerGroup("auto")]
internal sealed class EchoAutoCommandHandler(EvidenceStore evidence)
    : IZLinkSendHandler<EchoAutoCommand>
{
    public ValueTask HandleAsync(EchoAutoCommand message, ZLinkSendContext context, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add($"echo-command|variant=auto|id={message.CommandId}|value={message.Value}|content={context.ContentType}");
        return ValueTask.CompletedTask;
    }
}

[ZLinkHandlerGroup("attr")]
internal sealed class AttributeHandlers(EvidenceStore evidence)
{
    [ZLinkRequest(PacketName = "EchoAttr")]
    public EchoReply Request(EchoReq request, ZLinkRequestContext context, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add($"echo-request|variant=attr|value={request.Value}|content={context.ContentType}");
        return new EchoReply($"echo:{request.Value}", context.ContentType ?? "<null>");
    }

    [ZLinkSend(PacketName = "EchoAttrCommand")]
    public ValueTask Send(EchoCommand message, ZLinkSendContext context, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add($"echo-command|variant=attr|id={message.CommandId}|value={message.Value}|content={context.ContentType}");
        return ValueTask.CompletedTask;
    }
}

internal sealed class EchoManualRequestHandler(EvidenceStore evidence)
    : IZLinkRequestHandler<EchoManualReq, EchoReply>
{
    public ValueTask<EchoReply> HandleAsync(EchoManualReq request, ZLinkRequestContext context, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add($"echo-request|variant=manual|value={request.Value}|content={context.ContentType}");
        return ValueTask.FromResult(new EchoReply($"echo:{request.Value}", context.ContentType ?? "<null>"));
    }
}

internal sealed class EchoManualCommandHandler(EvidenceStore evidence)
    : IZLinkSendHandler<EchoManualCommand>
{
    public ValueTask HandleAsync(EchoManualCommand message, ZLinkSendContext context, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add($"echo-command|variant=manual|id={message.CommandId}|value={message.Value}|content={context.ContentType}");
        return ValueTask.CompletedTask;
    }
}

internal sealed class DuplicateEchoRequestHandler
    : IZLinkRequestHandler<EchoManualReq, EchoReply>
{
    public ValueTask<EchoReply> HandleAsync(EchoManualReq request, ZLinkRequestContext context, CancellationToken cancellationToken)
    {
        _ = context;
        cancellationToken.ThrowIfCancellationRequested();
        return ValueTask.FromResult(new EchoReply(request.Value, "duplicate"));
    }
}

File: framework/languages/node/e2e/RegistrationCodec/Server/Main/Handlers/registration-handlers.ts
import { Injectable } from '@nestjs/common';
import {
  zlinkRequestHandler,
  zlinkSendHandler
} from '@zlink-systems/nestjs';
import type {
  ZLinkRequestContext,
  ZLinkRequestHandler,
  ZLinkSendContext,
  ZLinkSendHandler
} from '@zlink-systems/framework';
import { PacketNames, type EchoCommand, type EchoReply, type EchoReq } from '../../../Shared/messages';
import { EvidenceStore } from '../Infrastructure/evidence-store';

@zlinkRequestHandler('auto', PacketNames.echoAuto)
@Injectable()
export class EchoAutoRequestHandler implements ZLinkRequestHandler<EchoReq, EchoReply> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(request: EchoReq, context: ZLinkRequestContext): Promise<EchoReply> {
    this.evidence.add(`echo-request|variant=auto|value=${request.value}|content=${context.contentType}`);
    return { value: `echo:${request.value}`, contentType: context.contentType ?? '<null>' };
  }
}

@zlinkSendHandler('auto', PacketNames.echoAutoCommand)
@Injectable()
export class EchoAutoCommandHandler implements ZLinkSendHandler<EchoCommand> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(message: EchoCommand, context: ZLinkSendContext): Promise<void> {
    this.evidence.add(`echo-command|variant=auto|id=${message.commandId}|value=${message.value}|content=${context.contentType}`);
  }
}

@zlinkRequestHandler('attr', PacketNames.echoAttr)
@Injectable()
export class EchoAttrRequestHandler implements ZLinkRequestHandler<EchoReq, EchoReply> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(request: EchoReq, context: ZLinkRequestContext): Promise<EchoReply> {
    this.evidence.add(`echo-request|variant=attr|value=${request.value}|content=${context.contentType}`);
    return { value: `echo:${request.value}`, contentType: context.contentType ?? '<null>' };
  }
}

@zlinkSendHandler('attr', PacketNames.echoAttrCommand)
@Injectable()
export class EchoAttrCommandHandler implements ZLinkSendHandler<EchoCommand> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(message: EchoCommand, context: ZLinkSendContext): Promise<void> {
    this.evidence.add(`echo-command|variant=attr|id=${message.commandId}|value=${message.value}|content=${context.contentType}`);
  }
}

@Injectable()
export class EchoManualRequestHandler implements ZLinkRequestHandler<EchoReq, EchoReply> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(request: EchoReq, context: ZLinkRequestContext): Promise<EchoReply> {
    this.evidence.add(`echo-request|variant=manual|value=${request.value}|content=${context.contentType}`);
    return { value: `echo:${request.value}`, contentType: context.contentType ?? '<null>' };
  }
}

@Injectable()
export class EchoManualCommandHandler implements ZLinkSendHandler<EchoCommand> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(message: EchoCommand, context: ZLinkSendContext): Promise<void> {
    this.evidence.add(`echo-command|variant=manual|id=${message.commandId}|value=${message.value}|content=${context.contentType}`);
  }
}

@Injectable()
export class DuplicateEchoRequestHandler implements ZLinkRequestHandler<EchoReq, EchoReply> {
  async handle(request: EchoReq): Promise<EchoReply> {
    return { value: request.value, contentType: 'duplicate' };
  }
}
